File: 2023/q5/q5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <limits>
#include "../../../../../usr/include/c++/11/bits/stream_iterator.h"

unsigned long long find_lowest_location(unsigned long long seed, const std::vector<std::vector<std::vector<std::string>>> &layers)
{

    unsigned long long curr = seed;
    for (const auto &layer : layers)
    {
        for (const auto &range_map : layer)
        {
            unsigned long long destination_range = std::stoll(range_map[0]);
            unsigned long long source_range = std::stoll(range_map[1]);
            unsigned long long length_range = std::stoll(range_map[2]);
            if (curr >= source_range && curr < source_range + length_range)
            {
                curr = (curr - source_range) + destination_range;
                break;
            }
        }
    }
    return curr;
}

int main(void)
{
    std::string filename = "q5.txt";
    std::ifstream FILE(filename);

    // Lines
    std::vector<std::vector<std::string>> lines;
    std::vector<std::string> curr;
    std::string line;

    while (std::getline(FILE, line))
    {
        if (line.empty())
        {
            lines.push_back(curr);
            curr.clear();
        }
        else
        {
            curr.push_back(line);
        }
    }
    if (!curr.empty())
    {
        lines.push_back(curr);
    }

    // Seeds
    std::vector<std::string> seeds;
    std::string seed_line = lines[0][0].substr(lines[0][0].find(": ") + 2);
    std::istringstream seed_stream(seed_line);
    std::string seed;
    while (seed_stream >> seed)
    {
        seeds.push_back(seed);
    }

    std::vector<std::vector<std::vector<std::string>>> layers;
    for (size_t i = 1; i < lines.size(); ++i)
    {
        std::vector<std::vector<std::string>> layer;
        for (size_t j = 1; j < lines[i].size(); ++j)
        {
            std::istringstream iss(lines[i][j]);
            std::vector<std::string> parts(std::istream_iterator<std::string>{iss}, std::istream_iterator<std::string>());
            layer.push_back(parts);
        }
        layers.push_back(layer);
    }

    unsigned long long ans = std::numeric_limits<unsigned long long>::max();

    for (size_t i = 0; i < seeds.size(); i += 2)
    {
        unsigned long long start = std::stoll(seeds[i]);
        unsigned long long length = std::stoll(seeds[i + 1]);
        for (unsigned long long seed = start; seed < start + length; ++seed)
        {
            ans = std::min(ans, find_lowest_location(seed, layers));
            std::cout << ans << "\n";
        }
    }

    std::cout << ans << std::endl;
}

File: 2023/q5/q5.py
import math
import threading

filename = "q5.txt"


def process_seed_range(start, length, layers, result):
    local_min = math.inf
    for seed in range(start, start + length):
        local_min = min(local_min, find_lowest_location(seed, layers))
        print(local_min)
    with result_lock:
        result[0] = min(result[0], local_min)


with open(filename, "r") as FILE:
    ans = 0

    # Process lines
    lines = []
    curr = []
    for line in FILE:
        if line == "\n":
            lines.append(curr)
            curr = []
        else:
            curr.append(line.strip())

    lines.append(curr)
    # Set up each instance
    # Each line is a list of strings

    # Part B Set Up Seeds
    seeds = lines[0][0].split(": ")[1].split(" ")

    layers = [lines[i][1:] for i in range(1, len(lines))]

    def layer_map_int(layer):
        ans = []
        for string in layer:
            ans.append([int(number) for number in string.split(" ")])

        return ans

    layers = list(map(layer_map_int, layers))

    ans = math.inf

    def find_lowest_location(seed, layers):
        # For each seed, walk thru each layer...
        location = math.inf

        curr = seed
        for layer in layers:
            # Check if curr is valid in this layer
            # If not in any range, the curr remains the same
            # print(f"Curr layer is {layer}")
            for range_map in layer:
                destination_range, source_range, length_range = range_map
                # print(destination_range, source_range, length_range)
                if curr >= source_range and curr < source_range + length_range:
                    """print(
                        f"Found in between {source_range} and {source_range + length_range}, destination_range {destination_range} and curr {curr}"
                    )
                    print(f"New curr to be {curr - source_range + destination_range}")
                    """
                    curr = (curr - source_range) + destination_range
                    break
            # print(curr)
        # print(f"Final location is {curr} for seed {seed}")
        return curr

    seeds = [(seeds[i], seeds[i + 1]) for i in range(0, len(seeds), 2)]
    print(seeds)
    result = [math.inf]  # Shared result
    result_lock = threading.Lock()  # Lock for updating the result
    threads = []

    for start, length in seeds:
        start = int(start)
        length = int(length)
        thread = threading.Thread(
            target=process_seed_range, args=(start, length, layers, result)
        )
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    ans = result[0]
    print(ans)
